### PR TITLE
SIL: Serialize @convention(c) thunks when referenced from an inlinable function

### PR DESCRIPTION
**Pull request by slavapestov**
Fixes <rdar://47719569>.

### CODE REVIEW

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/SIL/SILDeclRef.cpp [hunk 1]
```diff
@@ -521,6 +516,23 @@ IsSerialized_t SILDeclRef::isSerialized() const {
...
+  // local functions and closures, and can be referenced
+  // from serialized code.
+  if (isForeign) {
+    if (!getAbstractClosureExpr() &&
+        getDecl()->getDeclContext()->isTypeContext()) {
+      return IsNotSerialized;
+    }
+
+    return IsSerializable;
+  }
```
Comment (jrose-apple) [inline lib/SIL/SILDeclRef.cpp]: Does it hurt in practice to make them always serializable?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

File: lib/SIL/SILDeclRef.cpp [hunk 2]
```diff
@@ -444,83 +444,114 @@ bool SILDeclRef::isTransparent() const {
...
-    if (kind == SILDeclRef::Kind::DefaultArgGenerator) {
-      auto *afd = cast<AbstractFunctionDecl>(d);
-      switch (afd->getDefaultArgumentResilienceExpansion()) {
-      case ResilienceExpansion::Minimal:
-        return IsSerialized;
-      case ResilienceExpansion::Maximal:
-        return IsNotSerialized;
-      }
+
+    // Otherwise, ask the AST if we're inside an @inlinable context.
```
Comment (jrose-apple) [inline lib/SIL/SILDeclRef.cpp]: "Otherwise" compared to what?

File: lib/SIL/SILDeclRef.cpp [hunk 3]
```diff
@@ -444,83 +444,114 @@ bool SILDeclRef::isTransparent() const {
...
-      if (!scope.isPublic())
-        return IsNotSerialized;
-      if (nominal->isFormallyResilient())
-        return IsNotSerialized;
+  // Anything else that is not public is not serializable.
+  if (d->getEffectiveAccess() < AccessLevel::Public)
+    return IsNotSerialized;
+
+  // 'read' and 'modify' accessors synthesized on-demand are serialized if
+  // visible outside the module.
```
Comment (jrose-apple) [inline lib/SIL/SILDeclRef.cpp]: This seems wrong in a resilient world.

File: lib/SIL/SILDeclRef.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/SIL/SILDeclRef.cpp]: *Now* I think this could always be `IsSerializable` without causing any issues.

**Pull request merged**